[WIP] Audio Queue Manager

davidmroth opened the pull request:
Added an audio queue manager to fix Audio Player playback issue (were multiple audio stream played simultaneously). 

Known issues:
- [x] Playback error (multiple audio streams playing at once)
- [x] Audio didn't stop when keyword was detected when the "News Briefing" was playing
- [x] Stop, Pause, Play, Resume audio player not working
- [ ] Media playing causes false positive of keyword detection

davidmroth committed:
Audio Queue Manager

renekliment commented:
Thanks for the update.
Please:
- **DON'T** close / open new pull requests - you can just add commits to the existing and we can track changes; this way we don't see the changes (code and comments); when all issues are resolved, we can always squash the PR to be one commit; 
- `queue_audio` and `play_audio` should be also in `audio.py`, but renamed without the `_audio`
- do you need `import Queue` in `main.py`? 
- don't change indentation on/around lines you don't edit

Thanks!

davidmroth committed:
Audio Queue Manager

davidmroth commented:
Thanks for the feedback!  Already working on the suggested changes.

davidmroth commented:
FYI, the two open items are existing issues. I want to make sure my PR isn't held up as a result.

- Stop, Pause, Play, Resume audio player not working
- Audio player causes false positive of keyword detection (due to audio playing while active listening is on).
- 

I'll work on these items after the PR is deployed.

renekliment commented:
@davidmroth I see. Feel free to file an issue in the tracker for those if you want to.

davidmroth committed:
modified: main.py
	* Moving audio player subsystem and global components to seperate files

deleted: alexapi/audio.py
	* Renamed to alexapi/player.py

new file: alexapi/player.py
	* Dedicated Audio Player (seperate from Alexa voice response player) with queue handling
	* Implemented audio queueing
	* Moved queue_and_play and play (dropped the _queue; previously named queue_audio and play_audio repectively) into alexapi/player.py

new file: alexapi/player_state.py
	* Moved shared audio state variable to a dedicated file so they can be used globally

new file: alexapi/shared.py
	* Moved shared items to a dedicated file so they can be used globally
          - Configuration
	  - Debugging
          - GPIO

General:
	* TODO: More work needed

davidmroth committed:
modified: main.py
	* Renamed audio player functions

modified file: alexapi/player.py
	* Dedicated Audio Player (seperate from Alexa voice response player) with queue handling

modified file: alexapi/player_state.py
	* Minor update

General:
	* TODO: More work needed

davidmroth committed:
modified: main.py
	* Renamed audio player functions

modified file: alexapi/player.py
	* Dedicated Audio Player (seperate from Alexa voice response player) with queue handling

modified file: alexapi/player_state.py
	* Minor update

General:
	* TODO: More work needed

davidmroth committed:
Fixed merge

davidmroth commented:
modified: main.py
        * Renamed audio player functions
        * Moving audio player subsystem and global components to separate files
    
 modified file: alexapi/player.py
        * Dedicated Audio Player (separate from Alexa voice response player) with queue handling
        * Implemented audio queueing
          
 new file: alexapi/player_state.py
        * Moved shared audio state variable to a dedicated file so they can be used globally
              - Configuration
              - Debugging
              - GPIO

 deleted: alexapi/audio.py
        * Renamed to alexapi/player.py

renekliment commented on src/alexapi/player.py:
```diff
@@ -0,0 +1,302 @@
+	alexa_playback_progress_report_request=playback_progress_report_request
+	alexa_getnextitem = getnextitem
+
+"""
```
This line and down (I don't know how to select multiple lines):
Get rid of these encapsulations. Just create objects in `main.py` that call the stuff directly.

renekliment commented on src/alexapi/player.py:
```diff
@@ -0,0 +1,302 @@
+"""
+Setup
+"""
+def setup(playback_progress_report_request, getnextitem, tplaylist):	#TODO:I'm sure there is a better way
```
If you need to pass callbacks, do it via constructors ^, don't do this global stuff.

renekliment commented on src/alexapi/player_state.py:
```diff
@@ -0,0 +1,7 @@
+nav_token = ""
```
This whole thing should be in `player.py`. The state of the player is a part of it.

renekliment commented on src/alexapi/player.py:
```diff
@@ -0,0 +1,302 @@
+
+
+# Initialize players
+alexa_voice_response_player = AlexaVoiceResponsePlayer()
```
Do the object creation in `main.py`, where you can pass stuff via constructors.

renekliment commented on src/alexapi/player.py:
```diff
@@ -0,0 +1,302 @@
+
+		while True:
+			if media_playing is False and media_paused is False:
+				media_paused = False
```
This does nothing.

renekliment commented on src/alexapi/player.py:
```diff
@@ -0,0 +1,302 @@
+media_paused = False
+
+class AlexaVoiceResponsePlayer(object):
+	instance = None
```
Call this _vlc_, _vlcInstance_ or something. Just _instance_ sounds like it's a static class variable that holds the instance in case of a singleton design pattern.

renekliment commented on src/alexapi/player.py:
```diff
@@ -0,0 +1,302 @@
+class MediaPlayer(object):
+	q = deque() #A Thread safe queue used to play audio files play sequencially (FIFO)
+
+	instance = None
```
Same as above.

renekliment commented on src/alexapi/shared.py:
```diff
@@ -0,0 +1,37 @@
+import yaml
```
Get rid of this file. 
- `bcolors` should be in a separate file - [like this](https://github.com/renekliment/AlexaPiNG/blob/816f46d5901637da8ed1689b755b4f133b69aeda/src/alexapi/bcolors.py) - it is unfortunately not merged yet, but will, be, you can use that single file
- from `main.py`, pass the `config` via constructors - this is gonna be for example used when reading audio output device from config
- don't use global `debug` variable in the module - use `self.config['debug']`in the classes or something like that
- in `main.py` before passing `config` in the module (like [here](https://github.com/renekliment/AlexaPiNG/blob/816f46d5901637da8ed1689b755b4f133b69aeda/src/main.py), it's unfortunately not merged yet): 
```
if 'debug' not in config:
	config['debug'] = debug
``` 
- when you've got `debug` like this you don't need the command line parsing stuff
- put `import RPi.GPIO as GPIO` in `player.py` until #30 gets merged ... it's silly, but that's all we can do for now
- whoa, file gone

renekliment commented on src/main.py:
```diff
@@ -278,46 +235,48 @@ def process_response(r):
 						type_token = directive['payload']['adjustmentType']
 						if (type_token == 'relative'):
-							currVolume = currVolume + int(vol_token)
+							pstate.currVolume = pstate.currVolume + int(vol_token)
```
`pstate.currVolume += int(vol_token)`

renekliment reviewed (changes requested):
So this is quite the update!
I have a few requests, I sincerely hope they don't discourage you from contributing :) I'll be thrilled to merge this PR once it's resolved.

A few of them are our fault, because #30 isn't merged yet, but I hope you can use a few things from that PR that I've pointed out.

Apart from code comments, I have one question. From a very quick look it looks like there are two players - what's the reason for that? I thought that there's gonna be a single queue, but in `main.py` you call two distinct functions instead of `play_audio`.

Thank you!

renekliment commented:
So #36 finally got merged. Are you okay solving the merge conflict? If not, let me know and I'll do that after you solve the proposed changes.

Also, this means that your PR has to pass the tests (it's errors-only pylint test only so far) in order to get merged.

renekliment commented:
And #30 got merged, which might produce further conflicts. Do you need any help with that? What's your status on this? I'm planning a huge code style fix PR, which would make anything hard to merge, so I'd like to merge this before that. Thanks.

davidmroth commented:
Sorry, was busy with some other things. I'll get to work on this on tomorrow and shoot for having this done by Friday.

Thank you for your feedback and patience!

renekliment commented:
Awesome to hear that. Thank you!

renekliment commented:
@davidmroth 
Hey! I'm very sorry, but we have a lot of audio issues and I've decided not to wait and I have opened #68 to support playback_handlers and while I was at it I've implemented a queue there.

I've seen in your PR you have some additional features if I'm correct. Feel free to add those once #68 gets merged. Design opinions are also welcome, since I'm well aware that the current implementation is not ideal.

I've seen that you're working on the new API, which is awesome! This might also take off a bit from your shoulders so you can focus on the API if you want to. The thing I'd like to suggest is you stay up-to-date with our `master`, so you don't need to waste time doing things your way only to find out you have to rewrite it later, because it's already been resolved in `master`.

I'm sorry again and hope it's fine. I have to admit, I took a little inspiration from your PR, so it did help and I hope we'll see more contributions from you (such as the API), it's greatly appreciated!

renekliment commented:
P.S.: If you wanna _developer-talk_, join us at Gitter - link in the README.
